### PR TITLE
niftis2pam, pam2niftis Worflow

### PR DESCRIPTION
**Pull request by skoudoro**
This PR adds 2 utility command lines to convert nifti to pam5 and inverse.

**Todo:**
- [x] Add tests
- [x] Check `sphere`  management
- [x] Update doc

### CODE REVIEW

**Comment by arokem** [comment]
I haven't looked at this closely, but I am just here to make sure that this is done with an eye towards compliance with the BIDS DWI derivatives spec: https://github.com/bids-standard/bids-bep016/. In particular, things are evolving in https://github.com/bids-standard/bids-bep016/pull/5

**Comment by skoudoro** [comment]
Thank you for the lead, will look it to it. I was not aware to that

File: dipy/workflows/io.py [hunk 1]
```diff
@@ -205,3 +206,100 @@ def run(self, data_names, out_dir=''):
...
+
+    @classmethod
+    def get_short_name(cls):
+        return 'niftis_to_pam'
+
+    def run(self, peaks_dir_files, peaks_values_files, peaks_indices_files,
+            shm_files, gfa_files, out_dir='', out_pam="peaks.pam5"):
+        """Convert pam5 files to mutiple nifti files.
+
+        Parameters
```
Comment (gabknight) [inline dipy/workflows/io.py]: are all niftis mandatory in the pam format?

File: dipy/io/peaks.py [hunk 2]
```diff
@@ -199,28 +199,47 @@ def save_peaks(fname, pam, affine=None, verbose=False):
...
-
-        save_nifti(fname_dirs, pam_dirs, pam.affine)
-
-        save_nifti(fname_values, pam.peak_values.astype(np.float32),
-                   pam.affine)
-
-        save_nifti(fname_indices, pam.peak_indices, pam.affine)
-
-        save_nifti(fname_gfa, pam.gfa, pam.affine)
+def peaks_to_niftis(pam, fname_shm, fname_dirs, fname_values, fname_indices,
```
[the remark this answers is not in the document]
Comment (skoudoro) [reply dipy/io/peaks.py]: sounds good to me, I will update that

[the remark this answers is not in the document]
Comment (skoudoro) [reply dipy/io/peaks.py]: Thanks!

File: dipy/workflows/io.py (hunk 1 above)
Comment (skoudoro) [reply dipy/workflows/io.py]: yes!

**Comment by skoudoro** [comment]
Ok, I think I am done with this PR and it is ready for a review. I spent much more time than I expected on these new workflows.

**Comment by skoudoro** [comment]
Any comments @gabknight @arokem @Garyfallidis or others?

**Comment by jhlegarreta** [comment]
I have not followed this from the beginning, have not tested the workflow, and I grant the `PeaksAndMetrics` object is one that I have not used as extensively as others. Thus, I was not aware of DIPY's `pam` file format.

IMHO, the `pam` term may lead to confusion: there is actually a `pam` file format that I guess has little to do with our case. See [here](https://manpages.debian.org/stretch/netpbm/pam.5.en.html) or [here](http://netpbm.sourceforge.net/doc/pam.html).

Even if at first sight it fit well for the `PeaksAndMetrics` it may be worthwhile thinking of another extension to avoid any confusion.

Other than that, I'm always in favor of doing separate PRs when the nature of changes may advise to do so (in this case, changes to docstrings that are not necessarily a consequence of adding this workflow). It is in any case, it is great that these got fixed.

Despite my concerns/comments, please go ahead and merge this if you feel like it should be merged.

**Comment by Garyfallidis** [comment]
@skoudoro can you please rebase this?

File: dipy/reconst/eudx_direction_getter.pyx [hunk 3]
```diff
@@ -26,6 +26,15 @@ cdef class EuDXDirectionGetter(DirectionGetter):
         self.qa_thr = 0.0239
         self.ang_thr = 60
         self.total_weight = .5
+        self.sphere = None
```
Comment (gabknight) [inline dipy/reconst/eudx_direction_getter.pyx:29]: This seems to be the source of the test error. Could you move these out to the `PeaksAndMetrics` python class instead?

**Comment by Garyfallidis** [comment]
Ping!

**Comment by Garyfallidis** [comment]
We need to look back to this. @skoudoro ?

**Comment by skoudoro** [comment]
Ok, I think this old PR is ready. It would be great if you could have a look @gabknight or @arokem !

File: dipy/workflows/reconst.py [hunk 4]
```diff
@@ -746,10 +772,12 @@ def run(
         roi_radii=10,
         fa_thr=0.7,
         frf=None,
-        extract_pam_values=False,
-        sh_order=8,
-        odf_to_sh_order=8,
```
Comment (arokem) [inline dipy/workflows/reconst.py:751]: Was `odf_to_sh_order` not being used at all?

**Review by arokem — COMMENTED** [review]
Looks good. I am not sure that I understand what happened to `odf_to_sh_order` in some of these (see inline comment)

File: dipy/io/peaks.py [hunk 5]
```diff
@@ -40,10 +48,28 @@ def load_peaks(fname, *, verbose=False):
...
+    Parameters
+    ----------
+    fname : string
+        Filename of PAM5 file.
+    verbose : bool, optional
+        Print summary information about the loaded file.
+
+    Returns
+    -------
+    pam : PeaksAndMetrics object
```
Comment (jhlegarreta) [inline dipy/io/peaks.py:68]: Add description.

File: dipy/io/peaks.py [hunk 6]
```diff
@@ -214,20 +269,258 @@ def peaks_to_niftis(
...
+        Object holding peak_dirs, shm_coeffs and other attributes
+    fname_shm : str
+        Spherical Harmonics coefficients filename
+    fname_dirs : str
+        peaks direction filename
+    fname_values : str
+        peaks values filename
+    fname_indices : str
+        peaks indices filename
+    fname_gfa : str
```
Comment (jhlegarreta) [inline dipy/io/peaks.py:286]: , optional

File: dipy/io/peaks.py [hunk 7]
```diff
@@ -214,20 +269,258 @@ def peaks_to_niftis(
...
+    gfa : ndarray, optional
+        generalized FA volume
+    B : ndarray, optional
+        Matrix that transforms spherical harmonics to spherical function
+    qa : array, optional
+        Quantitative Anisotropy in each voxel.
+    odf : ndarray, optional
+        SH coefficients for the ODF spherical function
+    total_weight : float, optional
+    ang_thr : float, optional
```
Comment (jhlegarreta) [inline dipy/io/peaks.py:410]: Add a description to the above.

File: dipy/io/peaks.py [hunk 8]
```diff
@@ -214,20 +269,258 @@ def peaks_to_niftis(
...
+        Quantitative Anisotropy in each voxel.
+    odf : ndarray, optional
+        SH coefficients for the ODF spherical function
+    pam_file : str, optional
+        Filename of the desired pam file
+    npeaks : int, optional
+        Maximum number of peaks found (default 5 peaks).
+    generate_peaks_indices : bool, optional
+    total_weight : float, optional
+    ang_thr : float, optional
```
Comment (jhlegarreta) [inline dipy/io/peaks.py:489]: Add a description to the above.

File: dipy/workflows/reconst.py [hunk 9]
```diff
@@ -347,6 +380,14 @@ def run(
         save_metrics : variable string, optional
             List of metrics to save.
             Possible values: fa, ga, rgb, md, ad, rd, mode, tensor, evec, eval
+        nifti_tensor : bool, optional
+            Whether the tensor is saved in the standard Nifti format or in an
+            alternate format
+            that is used by other software (e.g., FSL): a
```
Comment (jhlegarreta) [inline dipy/workflows/reconst.py:386]: Unnecessary line break here.

File: dipy/workflows/reconst.py [hunk 10]
```diff
@@ -793,16 +821,21 @@ def run(
             (from the command line) or [15, 4, 4] from a Python script to be
             converted to float and multiplied by 10**-4 . If None
             the fiber response function will be computed automatically.
-        extract_pam_values : bool, optional
-            Save or not to save pam volumes as single nifti files.
-        sh_order : int, optional
+        sphere_name : string, optional
+            Sphere name on which to reconstruct the fODFs.
+        relative_peak_threshold : float, optional
+            Only return peaks greater than ``relative_peak_threshold * m``
```
Comment (jhlegarreta) [inline dipy/workflows/reconst.py:827]: Not sure how this will be rendered due to the asterisk.

**Review by jhlegarreta — COMMENTED** [review]
Have not followed everything; have only gone through the documentation.

Please use sentence case (capitalize first letter) and finish sentences with a period.

Does this PR change the API?

File: dipy/workflows/reconst.py (hunk 4 above)
Comment (skoudoro) [reply dipy/workflows/reconst.py:751]: No, it was not used, that's why I clean it up and clarify

File: dipy/workflows/reconst.py (hunk 10 above)
Comment (skoudoro) [reply dipy/workflows/reconst.py:827]: like this:

image

**Comment by skoudoro** [comment]
Thank you for your feedback @arokem and @jhlegarreta. I think I have fixed all the requested changes

**Review by jhlegarreta — COMMENTED** [review]
> Thank you for your feedback @arokem and @jhlegarreta. I think I have fixed all the requested changes

:+1:. So the API is changed here, even if only optional arguments are affected. Not sure the policy of a breaking change with respect to the planned minor version release, as this PR is included there.

**Comment by jhlegarreta** [comment]
> Here, the removed parameters fell in one of this category:

OK, unclear whether I see this in the changes made, but please good ahead.

**Comment by Garyfallidis** [comment]
Nice work @skoudoro

**Pull request merged**